Add functional form for classNames

rhdeck opened the pull request:
## Summary
Disconnect react-toastify from its SCSS classes by passing replacement className props as functions that contain context.

## Problem
If one wants to use a utility class framework (Tailwind!) 
1. The default classes keep interfering. 
2. Contextual information (what kind of toast is this?) is not available

## Solution 
Three props get changed to a function - or - string API:
1. `Toast.className`
2. `Toast.bodyClassName` (and `ToastContainer.toastClassName`)
3. `Toast.progressBarClassName` (and `ToastContainer.progressBarClassName`)

The string API works as it did before - class(es) added through this prop are mixed via `cx` with the toastify classes compiled from CSS

The functional API takes an arrow function (or `useCallback` return value) to set the string based on contextual information
```ts
(args: {type: 
  | 'info'
  | 'success'
  | 'warning'
  | 'error'
  | 'default'
  | 'dark'; defaultClassName: string; rtl: boolean;  position?: 
  | 'top-right'
  | 'top-center'
  | 'top-left'
  | 'bottom-right'
  | 'bottom-center'
  | 'bottom-left';}) => string
```

The `ToastContainer.className` also takes a functional form, though without the `type` prop:
```ts
(args: {defaultClassName: string; rtl: boolean;  position?: 
  | 'top-right'
  | 'top-center'
  | 'top-left'
  | 'bottom-right'
  | 'bottom-center'
  | 'bottom-left';}) => string
```

The functional API takes an arrow function (or `useCallback` return value) to set the string based on contextual information
```ts
(args: {defaultClassName: string; rtl: boolean;  position: 'top-right'
  | 'top-center'
  | 'top-left'
  | 'bottom-right'
  | 'bottom-center'
  | 'bottom-left';}) => string
```

This allows one to create code flows that determine which string to draw based on the type, and choose whether to mix in the default classes. 

### Example
A tailwind CSS/UI styled notification, without editiing SCSS.
```jsx
import { ToastContainer } from "@react-toastify";
import "react-toastify/dist/ReactToastify.css";
const contextClass = {
  success: "bg-blue-600",
  error: "bg-red-600",
  info: "bg-gray-600",
  warning: "bg-orange-400",
  default: "bg-indigo-600",
  dark: "bg-white-600 font-gray-300",
};
const App:FC = ()=>{
  return (
   <Main />
   <ToastContainer
      toastClassName={({ type }) => contextClass[type || "default"] + 
        " flex p-1 min-h-10 rounded-md justify-between overflow-hidden cursor-pointer"
      }
      bodyClassName={() => "text-sm font-white font-med block p-3"}
      position="bottom-left"
      autoClose={3000}
    />
  )
}
```
## Checklist:
- Added tests for functional form
- Implemented in production (distributed as [`@raydeck/react-toastify`](https://npmjs.com/package/@raydeck/react-toastify) until this gets merged)
- Doesn't break existing API (see tailwind sample above)


## Next steps post-merge
**DONE** Most likely other context args should get pulled in for a more powerful API. RTL? Position?

---
**Before submitting a pull request,** please make sure the following is done:

- [X] 1. Fork [the repository](https://github.com/fkhadra/react-toastify) and create your branch from `master`. 
- [X] 2. Run `yarn` in the repository root.
- [X] 3. If you've fixed a bug or added code that should be tested, add tests!
- [X] 4. Ensure the test suite passes (`yarn test`).
- [X] 5. Run `yarn start` to test your changes in the playground.
- [X] 6. Update the readme is needed
- [X] 7. Update the typescript definition is needed
- [X] 8. Format your code with [prettier](https://github.com/prettier/prettier) (`yarn prettier-all`).
- [X] 9. Make sure your code lints (`yarn lint:fix`).

**Learn more about contributing [here](https://github.com/fkhadra/react-toastify/blob/master/CONTRIBUTING.md)**

fkhadra commented:
Hey @rhdeck,

I'm currently testing your changes. I never used tailwind but I've always wanted to 😄. You just gave me a good excuse to do so. 

Thank you for the quality of the PR, really appreciate that.

fkhadra commented on src/components/ToastContainer.tsx:
```diff
@@ -24,18 +24,20 @@ export const ToastContainer: React.FC<ToastContainerProps> = props => {
-          ),
+          className:
+            typeof className === 'function'
+              ? className({})
```
@rhdeck shouldn't we pass the default className here ?

rhdeck replied:
Done

fkhadra commented on src/components/ToastContainer.tsx:
```diff
@@ -80,11 +82,11 @@ if (process.env.NODE_ENV !== 'production') {
     closeOnClick: PropTypes.bool,
     newestOnTop: PropTypes.bool,
-    className: PropTypes.string,
+    className: PropTypes.any, //oneOfType([PropTypes.func, PropTypes.string]),
```
I'll probably drop `prop-types` in a future release. So this is ok for me.

rhdeck replied:
Yes prop types are distressingly loose when we have good strict TS to work with

fkhadra commented on src/components/ProgressBar.tsx:
```diff
@@ -82,18 +82,20 @@ export function ProgressBar({
+    }
+  ];
+  const classNames =
+    typeof className === 'function'
```
May I ask to replace `typeof className === 'function'` by isFn(className)? Minify better

rhdeck replied:
Done

fkhadra commented on src/components/Toast.tsx:
```diff
@@ -36,18 +36,21 @@ export const Toast: React.FC<ToastProps> = props => {
+    }
+  ];
+  const cssClasses =
+    typeof className === 'function'
```
Can be replaced by `isFn`

rhdeck replied:
done

fkhadra commented on src/components/Toast.tsx:
```diff
@@ -67,14 +70,18 @@ export const Toast: React.FC<ToastProps> = props => {
           {...(props.in && { role: role })}
-          className={cx(`${DEFAULT.CSS_NAMESPACE}__toast-body`, bodyClassName)}
+          className={
+            typeof bodyClassName === 'function'
```
Can be replaced by `isFn`

rhdeck replied:
Done

fkhadra commented on src/components/ToastContainer.tsx:
```diff
@@ -24,18 +24,20 @@ export const ToastContainer: React.FC<ToastContainerProps> = props => {
-            parseClassName(className)
-          ),
+          className:
+            typeof className === 'function'
```
Can be replaced by `isFn`

rhdeck replied:
done

fkhadra commented on src/components/ToastPositioner.tsx:
```diff
@@ -14,7 +14,10 @@ export const ToastPositioner: React.FC<ToastPositionerProps> = ({
   return (
-    <div className={className!} style={style}>
+    <div
+      className={typeof className === 'function' ? className({}) : className}
```
This is not needed. The `className` is already computed by the `ToastContainer`

rhdeck replied:
A little type magic and toastpositioner takes a className as optional string again.

fkhadra reviewed (changes requested):
Currently, we pass the `type` and the `default classes`. Shouldn't we pass the position as well, what do you think ?

rhdeck commented:
Implemented requested fixes, updated the PR overview to reflect it, and made two additional improvements:

1. ToastContainer.className now has a more constrained argument list than Toast.className or ProgressBar.className (e.g. no type)
2. ToastClassName.type now restricts type to the ToastType string union, which should further reduce errors by users.

fkhadra reviewed (approved):
I'll merge it during the week end. Thanks for the work !

fkhadra commented:
Your PR landed in v6.1.0

michael-woodward-spok commented:
I think the docs still need to be updated, but this has made styling react-toastify so much easier! 🚀

fkhadra commented:
Yeah indeed. I'll try to update it this week.
